### PR TITLE
Implements prefilter and configures TinyGo unit tests

### PR DESCRIPTION
**Pull request by codefromthecrypt**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This implements the prefilter function and configures TinyGo unit tests via `make test-guest`. This runs the same unit tests without TinyGo on `make test`.

#### Special notes for your reviewer:

* This uses the name "prefilter" not "preFilter" because it matches the package name and helps clarify it is an operation name. We still use PreFilter when referring to framework.
* This implements prefilter node name results with a host callback for the same reason as the status reason: using a host callback allows the guest to control memory allocation. This uses a C-String approach to avoid complexity in marshalling. The same approach is used in http-wasm for header values.
* This intentionally does not do anything other design changes, such as caching state. We can do that in another PR.
* `tinygo test` uses a wasmtime by default, so the Makefile has been updated to switch to wazero.
* To avoid segfaults in TinyGo tests, this removes nottinygc as a strict dependency, and instead conventionally adds it when compiling plugins to wasm.
* This intentionally does not backfill more TinyGo unit tests because the change is already large.

#### Does this PR introduce a user-facing change?

NONE

#### What are the benchmark results of this change?

```benchstat
goos: darwin
goarch: arm64
pkg: sigs.k8s.io/kube-scheduler-wasm-extension/internal/e2e
                                                        │  before.txt  │              after.txt               │
                                                        │    sec/op    │    sec/op     vs base                │
PluginFilter/noop-wat/params:_small-12                    262.3n ±  2%   265.2n ±  1%       ~ (p=0.121 n=6)
PluginFilter/noop-wat/params:_real-12                     269.6n ±  2%   268.3n ±  1%       ~ (p=0.619 n=6)
PluginFilter/noop/params:_small-12                        333.1n ±  1%   328.0n ±  1%  -1.52% (p=0.002 n=6)
PluginFilter/noop/params:_real-12                         338.0n ±  2%   332.1n ±  3%  -1.75% (p=0.004 n=6)
PluginFilter/test/params:_small-12                        6.431µ ±  1%   6.438µ ±  0%       ~ (p=0.669 n=6)
PluginFilter/test/params:_real-12                         115.5µ ±  2%   116.8µ ±  0%  +1.14% (p=0.004 n=6)
PluginScore/noop-wat/params:_small-12                     260.6n ±  0%   259.7n ±  3%       ~ (p=0.056 n=6)
PluginScore/noop-wat/params:_real-12                      265.4n ±  0%   263.9n ±  0%  -0.55% (p=0.004 n=6)
PluginScore/noop/params:_small-12                         382.4n ±  1%   378.1n ±  0%  -1.15% (p=0.002 n=6)
PluginScore/noop/params:_real-12                          354.4n ±  0%   348.1n ±  0%  -1.76% (p=0.002 n=6)
PluginScore/test/params:_small-12                         3.668µ ±  1%   3.639µ ±  1%  -0.79% (p=0.002 n=6)
PluginScore/test/params:_real-12                          39.85µ ±  2%   39.16µ ± 12%       ~ (p=0.065 n=6)
PluginFilterAndScore/noop-wat/params:_small-12            379.8n ±  0%
PluginFilterAndScore/noop-wat/params:_real-12             390.6n ± 16%
PluginFilterAndScore/noop/params:_small-12                586.5n ±  5%
PluginFilterAndScore/noop/params:_real-12                 554.6n ±  6%
PluginFilterAndScore/test/params:_small-12                10.17µ ±  1%
PluginFilterAndScore/test/params:_real-12                 169.5µ ±  4%
PluginPreFilter/noop-wat/params:_small-12                                281.1n ±  3%
PluginPreFilter/noop-wat/params:_real-12                                 289.1n ±  1%
PluginPreFilter/noop/params:_small-12                                    347.2n ±  1%
PluginPreFilter/noop/params:_real-12                                     350.6n ±  0%
PluginPreFilter/test/params:_small-12                                    3.533µ ±  2%
PluginPreFilter/test/params:_real-12                                     40.56µ ±  9%
PluginPrefilterFilterAndScore/noop-wat/params:_small-12                  517.5n ±  1%
PluginPrefilterFilterAndScore/noop-wat/params:_real-12                   524.3n ±  3%
PluginPrefilterFilterAndScore/noop/params:_small-12                      788.3n ±  0%
PluginPrefilterFilterAndScore/noop/params:_real-12                       753.4n ±  0%
PluginPrefilterFilterAndScore/test/params:_small-12                      14.10µ ±  1%
PluginPrefilterFilterAndScore/test/params:_real-12                       218.8µ ±  1%
geomean                                                   1.436µ         1.429µ        -0.65%               ¹
```

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
fixed bench, as the prior commit didn't actually call prefilter. The below is higher because we don't yet cache the pod being scheduled. The next PR should show some dramatic difference

```
PluginFilterAndScore/test/params:_real-12                     169.5µ ±  4%
PluginPrefilterFilterAndScore/test/params:_real-12            218.8µ ±  1%
```

**Review by kerthcet — COMMENTED** [review]
/approve

File: scheduler/plugin/host.go [hunk 1]
```diff
@@ -71,12 +75,15 @@ type params struct {
 	// nodeName is used by guest.scoreFn
 	nodeName string
 
+	// resultNodeNames is returned by guest.prefilterFn
+	resultNodeNames []string
```
[the remark this answers is not in the document]
Comment (codefromthecrypt) [reply scheduler/plugin/host.go:79]: so this is a struct that has fields anything can use (e.g. the nodeName) I left it less explicit so I don't need to revisit the nodeName field. when something else uses nodeName (then need to say scoreOrSomethingElseNodeName). I'll rewrite the type docs to make the intent more clear.

File: guest/prefilter/cstring.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package prefilter
```
[the remark this answers is not in the document]
Comment (codefromthecrypt) [reply guest/prefilter/cstring.go:1]: gotcha.. is there a license plugin people use or should I grab one I use in another project?

File: scheduler/plugin/host.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (codefromthecrypt) [reply scheduler/plugin/host.go:89]: ok I added missing docs. PTAL!

**Pull request merged**